Implementing Andy Gio's mssql fix

roya0045 opened the pull request:
Implementing the suggested fix by Andy Gio in https://issues.qgis.org/issues/20592

## Description
This fix should allow the use of triggers on INSERT in MSSQL while maintaining the current behaviour.

To do so a temporary table is created in order to return the ID of the modified features instead of directly using OUTPUT to do so.

This fix has not been tested in my fork.
The prepare commit script has not been run since this is a minor change.

All credit should go to Mr. Andy Gio for this code that was provided in : https://issues.qgis.org/issues/20592.

roya0045 committed:
Implementing Andy Gio's mssql fix

Implementing the suggested fix by Andy Gio in https://issues.qgis.org/issues/20592

m-kuhn commented on src/providers/mssql/qgsmssqlprovider.cpp:
```diff
@@ -947,10 +952,14 @@ bool QgsMssqlProvider::addFeatures( QgsFeatureList &flist, Flags flags )
     if ( !( flags & QgsFeatureSink::FastInsert ) )
     {
-      statement += QStringLiteral( " OUTPUT inserted." ) + mFidColName;
+      statement += QStringLiteral( " OUTPUT inserted." ) + mFidColName + QStringLiteral( "  INTO @px " );
```
I don't know mysql very well. Should inserted and the column name be quoted (something like "inserted"."colname")?

roya0045 replied:
Quotations are not necessary but MSSQL likes when databases schema and tables are between bracket but the inserted table can be referenced without quotation and other such.

Here is an example with OUTPUT https://docs.microsoft.com/en-us/sql/t-sql/queries/output-clause-transact-sql?view=sql-server-2017#examples

m-kuhn replied:
Actually I think there should be a `QgsMssqlProvider::quotedIdentifier` method doing the same thing as the [`QUOTENAME`](https://docs.microsoft.com/en-us/sql/t-sql/functions/quotename-transact-sql?view=sql-server-2017) function. But that's out of scope for this pull request.

m-kuhn commented on src/providers/mssql/qgsmssqlprovider.cpp:
```diff
@@ -947,10 +952,14 @@ bool QgsMssqlProvider::addFeatures( QgsFeatureList &flist, Flags flags )
+    
+    if ( !( flags & QgsFeatureSink::FastInsert ) )
+    {    
+        statement += '; SELECT id FROM @px;';
```
```suggestion
         statement += QStringLiteral( "; SELECT id FROM @px;" );
```

m-kuhn commented:
Thank you for opening this pull request.
Did you compile and verify this fix? It doesn't compile in the current shape (see my suggestion) and I cannot verify the implementation because I don't have a mssql environment running here.

m-kuhn committed:
Update src/providers/mssql/qgsmssqlprovider.cpp

Wrapping the Select in QStringLiteral

Co-Authored-By: roya0045 <roya0045@users.noreply.github.com>

roya0045 commented:
I did not compile it. But I did some testing on a mssql environment prior to implementing this solution and the behaviour was the same as the current one without causing any issues when there are triggers on INSERT.

Thank you for the fix this was an oversight on my part.

roya0045 committed:
remove the empty statement

roya0045 committed:
fixing indentation

roya0045 committed:
Update qgsmssqlprovider.cpp

m-kuhn commented:
Looking good to me now.
Would be nice to get a pair of eyes from @nyalldawson @NathanW2 too.

roya0045 committed:
keeping up to date 1 (#1)

* New expression function sqlite_fetch_and_increment [FEATURE]

SQlite default values can only be applied on insert and not prefetched.
This makes it impossible to acquire an incremented primary key via AUTO_INCREMENT before creating the row in the database. Sidenote: with postgres, this works via the option "evaluate default values".

When adding new features with relations, it's really nice to be able to already add children for a parent, while the parents form is still open and hence the parent feature uncommitted.
To get around this limitation, this function can be used to manage an incrementing value (or multiple sequences) in a separate table on sqlite based formats like gpkg.

* Add test for sqlite_fetch_and_increment

* Add sqlite_fetch_and increment docs

* fixes #20673 : Correct mapFromSource method when reordering columns

* Remove template class

* Fix JSON formatting

* Processing in-place do not check validity when fixing geometries

Fixes #20812

* Drop the ugly hack in favor of sourceFlags

* Add sqlite3_database_unique_ptr::exec

A handy shortcut to quickly run commands on a sqlite database.

* Improve function help for sqlite_fetch_and_increment

* Check for proper error reporting in sqlite_fetch_and_increment

* Make geometry constraint configuration size fit the contents

* Toggle remove duplicate nodes when precision is changed

As soon as a precision is set, duplicate nodes are removed automatically. By
disabling the option and checking it in this scenario, this behavior is much
more transparent to the user.

* Don't hard-code black and gray for layer tree items' text color

* Add geometry check to API documentation

* Suppress api doc warnings for old classes

* Doxymenation for gap checks

* Skip more missing doc checks for old classes

* Some doc fixes

* Remove unneeded methods

* More doxygen

* Overlap check documentation

* Add geometry check classes to analysis group

* Even more docs

* Further doc improvements

* Remaining docs

* Make travis happy

* Optimize space on CRS selection tab

* Make travis even happier

* Make more space available on CRS selection dialog

* Limit size of projection preview map

* Revert QgsBlockingNetworkRequest

Too many issues... I'm unsure if this is even possible now...

* Tame doxygen

* debian packaging: include qgis.processing and qgis.testing (fixes #20856)

* [FEATURE] A new grayscale theme, "blend of gray"

* Missing @cond

* Use QGIS user manual for GDAL algorithms help files

because they are available for most of the algs and wouldn't be used otherwise. Also when someone hits algA help button, he should be given the algA help and not the main gdal utility doc.

* Only add option list widget styling for default theme
(This allows customized styling for non-default themes,
fixes white background on white text issue et cie)

* Blend of Gray theme fixes:
- white background on white text options/project/etc. panel list widget
- unwanted light gray tool button's menu button

* Additional Blend of Gray theme fixes:
- proper colors for selected item in welcome screen
- fix toolbar border pixel shift

* Add fast truncate implementation for memory provider

* Be theme-friendly, enforce borderless close button in message bar

* Added grass76 support

Added grass76 support; search for newer versions first

* Optimize alpha value for legend text color

m-kuhn commented:
No reason to merge master back into this pull request as long as there are no merge conflicts ;)

nyalldawson commented:
I'll test this next week when I'm back in the office and have access to a SQL Server instance

roya0045 commented:
@m-kuhn Nah I'm just not too used to github and wanted to keep my branch up to date for the other project and it just added this to this pull since it was my master, I should have branched it in restrospective.

m-kuhn commented:
I see, it's also nothing wrong with it, just wanted to let you know that it's unrequired, all good :)

roya0045 committed:
Revert "keeping up to date 1 (#1)" (#4)

This reverts commit 0accce84573ef143270deed132bcf604d7180d34.

roya0045 commented:
I'll reopen it it a new request I'm breaking my own repo too  much.
